### PR TITLE
Update dashboard JSON and fix dashboard example

### PR DESCRIPTION
**Pull request by petervdonovan**
~~This PR makes the following changes:~~
- ~~Update the dashboard to include fixes for objectively wrong data that used to appear in the reaction execution timelines. These fixes come from the mixed-criticality dashboard.~~
- ~~Rework the time series panel which is now called "Maximum lag by logical time." Notably, this changes it so that the x-axis is logical time, not physical time. When I was working on the mixed-criticality examples, I found that this make the panel more interpretable. This improvement comes from the mixed-criticality dashboard.~~
- ~~Remove the panel called "Logical time by reactor," which we have discussed on Zulip and which we seem to have agreed is uninformative and difficult to understand. This improvement also comes from the mixed-criticality dashboard.~~
- ~~Update the dashboard with descriptions to clarify the basic meaning of certain visualizations. This is a response to a comment from this morning that one of our users wanted documentation for the dashboard.~~
- ~~Change some variable names in the example programs to avoid naming collisions. A more robust and general fix that uses IDs to distinguish between different programs that use the same FQNs is left for a future PR.~~
- ~~Update the C example so that deadline violations sometimes occur. This is done simply to sanity-check that the deadline violation metrics do work.~~

**EDIT: see CHANGELOG.md for the updated changelist.**

I tested these changes locally, and they seem to work on my machine. I used LFC commit `2948f1ff8b6a95bdce678ce515c6eea3fa5a4999`.

The JSON of this version of the dashboard should work with the SDV demo. It does not contain uids that will be invalid. I believe that it can safely be copied over to that repo.

### CODE REVIEW

File: InfluxPublisher/README.md [hunk 1]
```diff
@@ -6,15 +6,15 @@ This folder contains Lingua Franca reactors that publish to InfluxDB via a socke
...
 
-    ```bash
-    docker compose up
-    ```
+   ```bash
+   docker compose up
+   ```
 
-    Open Grafana on `http://0.0.0.0:3000` and open the dashboard "Telegraf Test Publishers".
+   Open Grafana on `http://0.0.0.0:3000` and open the dashboard "Telegraf Test Publishers".
```
Comment (lhstrh) [inline InfluxPublisher/README.md]: Iwai-san reported:
> Open Grafana on http://0.0.0.0:3000 and open the dashboard "Telegraf Test Publishers".
-> http://localhost:3000

Comment (petervdonovan) [reply InfluxPublisher/README.md]: OK, done

File: InfluxPublisher/README.md [hunk 2]
```diff
@@ -6,15 +6,15 @@ This folder contains Lingua Franca reactors that publish to InfluxDB via a socke
...
 
-    ```bash
-    docker compose up
-    ```
+   ```bash
+   docker compose up
+   ```
 
-    Open Grafana on `http://0.0.0.0:3000` and open the dashboard "Telegraf Test Publishers".
+   Open Grafana on `http://localhost:3000` and open the dashboard "Telegraf Test Publishers".
```
Comment (elgeeko1) [inline InfluxPublisher/README.md]: Curious why this change?

File: InfluxPublisher/README.md [hunk 3]
```diff
@@ -6,15 +6,15 @@ This folder contains Lingua Franca reactors that publish to InfluxDB via a socke
 
 1. Compile the LF example applications:
 
-    ```bash
-    cd c && lfc src/InfluxPublisher.lf
-    cd py && lfc src/InfluxPublisher.lf
-    ```
+   ```bash
+   cd c && lfc src/InfluxPublisher.lf; cd ..
```
Comment (elgeeko1) [inline InfluxPublisher/README.md]: Ah, good cathch. I prefer `(cd c && lfc src/InfluxPublisher.lf)` since it's easier to combine with other commands, but just food for thought.

File: InfluxPublisher/c/src/InfluxPublisher.lf [hunk 4]
```diff
@@ -117,7 +123,7 @@ reactor SineWave(sample_rate_ms = 100 ms) {
 }
 
 main reactor {
-    publisher = new InfluxLinePublisher()
-    producer = new SineWave()
-    producer.measurement -> publisher.line
+    c_publisher = new InfluxLinePublisher()
```
Comment (elgeeko1) [inline InfluxPublisher/c/src/InfluxPublisher.lf:120]: Curious why this change? It makes sense from a naming standpoint, but wanted to know if there was another reason.

**Review by elgeeko1 — CHANGES_REQUESTED** [review]
One change to request here. The test publishers are designed to be templates that can be used for user-level publishing to InfluxDB. I think by adding deadlines we're mixing templates and examples, where the deadlines make sense only for examples, and don't make sense for a user who wants to have a template for how to publish to InfluxDB.

Instead, could we create a separate `examples` folder that has a publisher with deadline misses?

**Comment by elgeeko1** [comment]
I'm looking at the updated tracing dashboard now.

1. The tooltips are *very* helpful, as are the more detailed names for the panels.
2. The graph "Maximum lag time by logical time", I'm not sure I understand what is the meaning of lag in logical time with respect to the logical clock? I'm used to thinking of lag between logical time and physical time. Can you help me understand why this metric matters? (Also: the logarithmic scale is very clean)
3. I notice the update rate is now 1s, which we support with our latest `grafana.ini` configuration, but I'm surprised it works in the tracing dashboard. Can the Influx queries keep up with this? What did you learn when moving from a 5s update to a 1s update?
4. As I'm looking at the WCET diagram, I'm thinking of a new feature which is to show execution time as a percentage of total execution time of the application during the window, a proxy for CPU utilization. This is a feature request and not something we should build now.
5. Reaction executions, if I read this correctly, is showing a really valuable view. The behavior planner for example 
image shows some reactions executing nonstop. Is this what you would expect?
6. See the graph below. I see only green, what does this mean?
image
7. I see no deadline misses for the AVP demo, is this what you would expect?

I think it would be helpful to review the AVP demo together with you.

File: InfluxPublisher/README.md (hunk 2 above)
Comment (petervdonovan) [reply InfluxPublisher/README.md]: See https://github.com/xronos-inc/xronos-dashboard/pull/3#discussion_r1698997833. That's all I know.

File: InfluxPublisher/c/src/InfluxPublisher.lf (hunk 4 above)
Comment (petervdonovan) [reply InfluxPublisher/c/src/InfluxPublisher.lf:120]: This is the workaround that we discussed on Wednesday morning. It is this change (described above):
> Change some variable names in the example programs to avoid naming collisions. A more robust and general fix that uses IDs to distinguish between different programs that use the same FQNs is left for a future PR.

**Comment by petervdonovan** [comment]
Thanks for your comments. I'll look at items 2 and 3 tomorrow morning. For item 2, I think I need to clarify the "description" and edit the title so that it is clear that the lag is physical time minus logical time, as usual. For item 3, probably you are right, the refresh rate should be reverted back to 5s; the only difference is how pleasant it is to look at.

Item 4 I will keep in mind for later.

For item 6, I'll take another look at the "description" for that panel. It should explain that higher opacity corresponds to a later logical time.

Items 5 and 7 I am not sure about. I agree that it might be good to look at it together.

File: InfluxPublisher/README.md (hunk 3 above)
Comment (petervdonovan) [reply InfluxPublisher/README.md]: Sounds good, README.md has been updated with the improved Bash commands.

**Comment by petervdonovan** [comment]
> Instead, could we create a separate examples folder that has a publisher with deadline misses?

Yep, that is the better way. Done in [2944bdb](https://github.com/xronos-inc/xronos-dashboard/pull/3/commits/2944bdb7a55d203032242dd7e2586f27e7e3c81e).

**Comment by petervdonovan** [comment]
@elgeeko1 @cmnrd I think this PR is once again ready for review. Thanks for your patience as it turned out I did not do due diligence the first time around. I think that the bugs that were apparent in `Multi_Tread.lf` and `Single_Tread.lf` have been ironed out.

File: CHANGELOG.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+## [8ebc9b5](https://github.com/xronos-inc/xronos-dashboard/tree/8ebc9b504e602259b7ccfe64fc43a0a0e206c573) (2024-08-05)
+
+### Bugfixes
+
+- Update the dashboard to include fixes for incorrect end times and dropped reaction executions that used to appear in the reaction execution timelines.
+  - This also changes the panel titled "Reaction executions (opacity indicates logical time and microstep)" to use opacity to indicate logical time and microstep instead of using color to indicate logical time. This can make the panel difficult to read without zooming in, but for technical reasons, it is necessary in order to enable this bugfix.
+  - Example 1:
+    - Before this bugfix, reaction executions are often dropped or reported as having the wrong execution times, seemingly at random:
+      ![after](documentation/images/changelog/2024-08-05/multi_tread_master.png)
```
Comment (petervdonovan) [inline CHANGELOG.md]:
```suggestion
      ![before](documentation/images/changelog/2024-08-05/multi_tread_master.png)
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    - Before this bugfix, reaction executions are often dropped or reported as having the wrong execution times, seemingly at random:
+      ![before](docs/images/changelog/2024-08-05/multi_tread_master.png)
+    - After this bugfix, reaction executions are reported correctly, clearly showing that we periodically execute a 50-millisecond reaction in reactor `a` followed by short-running reactions in the other reactors:
+      ![after](docs/images/changelog/2024-08-05/multi_tread_corrected.png)
+  - Example 2 is similar to example 1:
+    - Before:
+      ![before](docs/images/changelog/2024-08-05/single_tread_master.png)
+    - After:
+      ![after](docs/images/changelog/2024-08-05/single_tread_corrected.png)
+- Fix a bug involving unclear and non-unique names that made the "Reaction WCET" panel difficult to interpret. This panel now reports distinct data for each reaction and does not aggregate data across distinct reactions in the same process.
```
Comment (cmnrd) [inline CHANGELOG.md]:
```suggestion
- Fix a bug involving unclear and non-unique names that made the "Reaction WCET" panel difficult to interpret. This panel now reports distinct data for each reaction and does not aggregate data across distinct reactions in the same federate.
```

**Review by cmnrd — APPROVED** [review]
I run this again and can confirm that this indeed fixes the bugs that we observed.

Should we move the InfluxPublisher example also to examples and perhaps give it a more descriptive name (e.g. SineWaveGenerator)?

**Comment by rcakella** [comment]
@elgeeko1, I tried this today and it looks good to me.
